src/darwin/Framework/CHIP/CHIPDevice.mm

```diff
@@ -173,6 +202,898 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
+        });
+}
+
+class NSObjectCommandCallback final : public app::CommandSender::Callback {
+public:
+    using OnSuccessCallbackType
+        = std::function<void(const app::ConcreteCommandPath &, const app::StatusIB &, const NSObjectData &)>;
+    using OnErrorCallbackType = std::function<void(CHIP_ERROR aError)>;
+    using OnDoneCallbackType = std::function<void(app::CommandSender * commandSender)>;
+
+    /*
+     * Constructor that takes in success, failure and onDone callbacks.
+     *
+     * The latter can be provided later through the SetOnDoneCallback below in cases where the
+     * TypedCommandCallback object needs to be created first before it can be passed in as a closure
+     * into a hypothetical OnDoneCallback function.
+     */
+    NSObjectCommandCallback(chip::ClusterId clusterId, chip::CommandId commandId, OnSuccessCallbackType aOnSuccess,
+        OnErrorCallbackType aOnError, OnDoneCallbackType aOnDone = {})
+        : mOnSuccess(aOnSuccess)
+        , mOnError(aOnError)
+        , mOnDone(aOnDone)
+        , mClusterId(clusterId)
+        , mCommandId(commandId)
+    {
+    }
+
+    void SetOnDoneCallback(OnDoneCallbackType callback) { mOnDone = callback; }
+
+private:
+    void OnResponse(app::CommandSender * apCommandSender, const app::ConcreteCommandPath & aCommandPath,
+        const app::StatusIB & aStatus, TLV::TLVReader * aReader) override;
+
+    void OnError(const app::CommandSender * apCommandSender, CHIP_ERROR aError) override { mOnError(aError); }
+
+    void OnDone(app::CommandSender * apCommandSender) override { mOnDone(apCommandSender); }
+
+    OnSuccessCallbackType mOnSuccess;
+    OnErrorCallbackType mOnError;
+    OnDoneCallbackType mOnDone;
+    chip::ClusterId mClusterId;
+    chip::CommandId mCommandId;
+};
+
+void NSObjectCommandCallback::OnResponse(app::CommandSender * apCommandSender, const app::ConcreteCommandPath & aCommandPath,
+    const app::StatusIB & aStatus, TLV::TLVReader * aReader)
+{
+    NSObjectData response;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    //
+    // Validate that the data response we received matches what we expect in terms of its cluster and command IDs.
+    //
+    VerifyOrExit(aCommandPath.mClusterId == mClusterId && aCommandPath.mCommandId == mCommandId, err = CHIP_ERROR_SCHEMA_MISMATCH);
+
+    if (aReader != nullptr) {
+        err = app::DataModel::Decode(*aReader, response);
+        SuccessOrExit(err);
+    }
+
+    mOnSuccess(aCommandPath, aStatus, response);
+
+exit:
+    if (err != CHIP_NO_ERROR) {
+        mOnError(err);
+    }
+}
+
+- (void)invokeCommandWithEndpointId:(NSUInteger)endpointId
+                          clusterId:(NSUInteger)clusterId
+                          commandId:(NSUInteger)commandId
+                      commandFields:(id)commandFields
+                        clientQueue:(dispatch_queue_t)clientQueue
+                         completion:(CHIPDeviceResponseHandler)completion
+{
+    new NSObjectAttributeCallbackBridge(
```

Comment:
Why is this using an _attribute_ callback bridge for commands?

Reply:
Because underlying mechanism will work in the same way to pass back the response data.

Reply:
My point is that perhaps the bridge should be renamed....